### ISSUE TITLE
[Feature] Nomad constraint for consul maint mode

### ISSUE THREAD

**Issue by jshaw86** [hashicorp/nomad#2283]
We would like there to be a scheduling constraint for when consul is in maintenance mode. A specific nomad client may get into a state where it can accept jobs but consul is in maintenance mode.

The affect then is if there are routing services like Fabio depending on the consul health checks the job is never routed to and considered down.

**Comment by dadgar** [comment]
I think if we fingerprint this you could add a constraint to avoid nodes in maintenance mode.

**Comment by a86c6f7964** [comment]
Would that move things off of that node? Seems like this could be taken care of with putting nomad into drain mode if consul is in maintenance mode.

**Comment by dadgar** [comment]
@a86c6f7964 it wouldn't automatically move them off, but if the job was rerun it would or if, as you said, you put Nomad in drain mode. Seems like the right thing to do.

Actually be putting Nomad into drain mode you wouldn't need the constraint at all.

**Comment by jshaw86** [comment]
@a86c6f7964 

> Seems like this could be taken care of with putting nomad into drain mode if consul is in maintenance mode.

You are correct accept what if you bring nomad out of drain but consul remains in maint? The current behavior is nomad will schedule to the node even with consul in maint causing all the consul infrastructure to ignore those containers.

Ideally this never happens but we've found consul to be non-deterministic when bringing out of maint mode.

**Comment by a86c6f7964** [comment]
In theory if there was a config to put nomad into drain when consul is set to maintenance then you would not be able to pull nomad out of drain while consul is in maintenance.

**Comment by dadgar** [comment]
@jshaw86 The mechanism to block Nomad from scheduling on that node is marking it as drain. I don't think we would want to special any other behavior. So maybe this is more of a Consul issue if you are saying it is non-determinstic?

**Comment by jshaw86** [comment]
@a86c6f7964 yes in theory not in practice currently :).

**Comment by jshaw86** [comment]
@dadgar yea maybe so, it would be nice to have something in nomad to prevent this from happening just to eliminate the potential of getting into this state but I can try to file an issue with consul.

I know you guys are trying to keep nomad and consul un coupled so not sure what the best approach would be if you were to do it.

**Comment by dadgar** [comment]
@jshaw86 I think what makes the most sense is to add an option in the client config that when Nomad is put into and out of drain mode it toggles the appropriate maintenance mode on Consul

**Comment by pierreca** [comment]
nomad drain toggling maintenance mode in consul would be good but not sufficient, as consul maintenance can be enabled without going through nomad drain. 

fingerprinting is an interesting route but you could also consider registering a node health check with consul and when that health check fails because of maintenance mode maybe drain it.

Because consul can also be used for service discovery, and services aren't advertised in maintenance mode, it seems reasonable for nomad to consider this allocation as failed or lost (or maybe another state, like "unusable") and reschedule it on another node.

Continuing to schedule, or considering an allocation healthy, when maintenance mode is enabled on that node (and therefore its health checks are failing) lead to broken or undiscoverable service instances.
